urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+          // Not a candidate, and can never become one.
+          equalities.clear();
+          lastequality = null;
+          safeDiff = thisDiff;
+        }
+        post_ins = post_del = false;
+      } else {
+        // An insertion or deletion.
+        if (thisDiff.operation == Operation.DELETE) {
+          post_del = true;
+        } else {
+          post_ins = true;
+        }
+        /*
+         * Five types to be split:
+         * <ins>A</ins><del>B</del>XY<ins>C</ins><del>D</del>
+         * <ins>A</ins>X<ins>C</ins><del>D</del>
+         * <ins>A</ins><del>B</del>X<ins>C</ins>
+         * <ins>A</del>X<ins>C</ins><del>D</del>
+         * <ins>A</ins><del>B</del>X<del>C</del>
+         */
+        if (lastequality != null
+            && ((pre_ins && pre_del && post_ins && post_del)
+                || ((lastequality.length() < Diff_EditCost / 2)
+                    && ((pre_ins ? 1 : 0) + (pre_del ? 1 : 0)
+                        + (post_ins ? 1 : 0) + (post_del ? 1 : 0)) == 3))) {
+          //System.out.println("Splitting: '" + lastequality + "'");
+          // Walk back to offending equality.
+          while (thisDiff != equalities.lastElement()) {
+            thisDiff = pointer.previous();
+          }
+          pointer.next();
+
+          // Replace equality with a delete.
+          pointer.set(new Diff(Operation.DELETE, lastequality));
+          // Insert a corresponding an insert.
+          pointer.add(thisDiff = new Diff(Operation.INSERT, lastequality));
+
+          equalities.pop();  // Throw away the equality we just deleted.
+          lastequality = null;
+          if (pre_ins && pre_del) {
+            // No changes made which could affect previous entry, keep going.
+            post_ins = post_del = true;
+            equalities.clear();
+            safeDiff = thisDiff;
+          } else {
+            if (!equalities.empty()) {
+              // Throw away the previous equality (it needs to be reevaluated).
+              equalities.pop();
+            }
+            if (equalities.empty()) {
+              // There are no previous questionable equalities,
+              // walk back to the last known safe diff.
+              thisDiff = safeDiff;
+            } else {
+              // There is an equality we can fall back to.
+              thisDiff = equalities.lastElement();
+            }
+            while (thisDiff != pointer.previous()) {
+              // Intentionally empty loop.
+            }
+            post_ins = post_del = false;
+          }
+
+          changes = true;
+        }
+      }
+      thisDiff = pointer.hasNext() ? pointer.next() : null;
+    }
+
+    if (changes) {
+      diff_cleanupMerge(diffs);
+    }
+  }
+
+  /**
+   * Reorder and merge like edit sections.  Merge equalities.
+   * Any edit section can move as long as it doesn't cross an equality.
+   * @param diffs LinkedList of Diff objects.
+   */
+  public void diff_cleanupMerge(LinkedList<Diff> diffs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 29. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1314-1357 into a separate method.